components/script/dom/mediasession.rs

```diff
@@ -0,0 +1,146 @@
+
+impl MediaSession {
+    #[allow(unrooted_must_root)]
+    fn new_inherited(browsing_context_id: TopLevelBrowsingContextId) -> MediaSession {
+        let media_session = MediaSession {
+            reflector_: Reflector::new(),
+            metadata: Default::default(),
+            playback_state: DomRefCell::new(MediaSessionPlaybackState::None),
+            action_handlers: DomRefCell::new(HashMap::new()),
+            media_instance: Default::default(),
+        };
+        ScriptThread::register_media_session(&media_session, browsing_context_id);
+        media_session
+    }
+
+    pub fn new(window: &Window) -> DomRoot<MediaSession> {
+        let browsing_context_id = window.window_proxy().top_level_browsing_context_id();
+        reflect_dom_object(
+            Box::new(MediaSession::new_inherited(browsing_context_id)),
+            window,
+            MediaSessionBinding::Wrap,
+        )
+    }
+
+    pub fn handle_action(&self, action: MediaSessionActionType) {
+        println!("HANDLE ACTION {:?}", action);
+        if let Some(handler) = self.action_handlers.borrow().get(&action) {
+            if handler.Call__(ExceptionHandling::Report).is_err() {
+                warn!("Error calling MediaSessionActionHandler callback");
+            }
+            return;
+        }
+        // TODO default action.
+    }
+
+    pub fn send_event(&self, event: MediaSessionEvent) {
+        let global = self.global();
+        let window = global.as_window();
+        window.send_to_embedder(EmbedderMsg::MediaSessionEvent(event));
+    }
+}
+
+impl MediaSessionMethods for MediaSession {
+    /// https://w3c.github.io/mediasession/#dom-mediasession-metadata
+    fn GetMetadata(&self) -> Option<DomRoot<MediaMetadata>> {
+        self.metadata.get()
+    }
+
+    /// https://w3c.github.io/mediasession/#dom-mediasession-metadata
+    fn SetMetadata(&self, metadata: Option<&MediaMetadata>) {
+        if let Some(ref metadata) = metadata {
+            metadata.set_session(self);
+        }
+        self.metadata.set(metadata);
+    }
+
+    /// https://w3c.github.io/mediasession/#dom-mediasession-playbackstate
+    fn PlaybackState(&self) -> MediaSessionPlaybackState {
+        *self.playback_state.borrow()
+    }
+
+    /// https://w3c.github.io/mediasession/#dom-mediasession-playbackstate
+    fn SetPlaybackState(&self, state: MediaSessionPlaybackState) {
+        *self.playback_state.borrow_mut() = state;
+    }
+
+    /// https://w3c.github.io/mediasession/#update-action-handler-algorithm
+    fn SetActionHandler(
+        &self,
+        action: MediaSessionAction,
+        handler: Option<Rc<MediaSessionActionHandler>>,
+    ) {
+        match handler {
+            Some(handler) => self
+                .action_handlers
+                .borrow_mut()
+                .insert(action.into(), handler.clone()),
+            None => self.action_handlers.borrow_mut().remove(&action.into()),
+        };
+    }
+}
+
+impl Drop for MediaSession {
```

Comment:
It might be worth revisiting this in the light of https://github.com/servo/servo/issues/24483